drivers/auxdisplay/auxdisplay_serlcd.c

```diff
@@ -0,0 +1,398 @@
+	const struct auxdisplay_capabilities capabilities = config->capabilities;
+	const uint16_t columns = capabilities.columns;
+	const uint16_t rows = capabilities.rows;
+	struct auxdisplay_serlcd_data *data = dev->data;
+
+	if (type == AUXDISPLAY_POSITION_ABSOLUTE) {
+		/*
+		 * shortcut for (0,0) position
+		 */
+		if (x == 0 && y == 0) {
+			data->cursor_x = x;
+			data->cursor_y = y;
+			return auxdisplay_serlcd_send_special_command(dev,
+								      SERLCD_SPECIAL_RETURN_HOME);
+		}
+
+		/*
+		 * bounds checking
+		 */
+		if (x < 0 || x >= columns) {
+			return -EINVAL;
+		}
+		if (y < 0 || y >= rows) {
+			return -EINVAL;
+		}
+		const uint8_t column = x;
+		const uint8_t row = y;
+
+		data->cursor_x = column;
+		data->cursor_y = row;
+
+		const uint8_t cursor_address = column + row_offsets[row];
+
+		return auxdisplay_serlcd_send_special_command(
+			dev, SERLCD_SPECIAL_SET_DD_RAM_ADDRESS | cursor_address);
+
+	} else if (type == AUXDISPLAY_POSITION_RELATIVE) {
+		/*
+		 * clip relative move to display dimensions
+		 */
+		const int new_x = (data->cursor_x + x) % columns;
+		const int new_y = (data->cursor_y + y + x / columns) % rows;
+		const uint16_t column = new_x < 0 ? new_x + columns : new_x;
+		const uint16_t row = new_y < 0 ? new_y + rows : new_y;
+
+		data->cursor_x = column;
+		data->cursor_y = row;
+
+		const uint8_t cursor_address = column + row_offsets[row];
+
+		return auxdisplay_serlcd_send_special_command(
+			dev, SERLCD_SPECIAL_SET_DD_RAM_ADDRESS | cursor_address);
+	}
+
+	/*
+	 * other types of movement are not implemented/supported
+	 */
+	return -ENOSYS;
+}
+
+static int auxdisplay_serlcd_cursor_position_get(const struct device *dev, int16_t *x, int16_t *y)
+{
+	const struct auxdisplay_serlcd_data *data = dev->data;
+
+	*x = (int16_t)data->cursor_x;
+	*y = (int16_t)data->cursor_y;
+
+	return 0;
+}
+
+static int auxdisplay_serlcd_capabilities_get(const struct device *dev,
+					      struct auxdisplay_capabilities *capabilities)
+{
+	const struct auxdisplay_serlcd_config *config = dev->config;
+
+	memcpy(capabilities, &config->capabilities, sizeof(struct auxdisplay_capabilities));
+
+	return 0;
+}
+
+static int auxdisplay_serlcd_clear(const struct device *dev)
+{
+	int rc = auxdisplay_serlcd_send_command(dev, SERLCD_COMMAND_CLEAR);
+
+	/*
+	 * clearing takes more time
+	 */
+	k_sleep(K_MSEC(10));
+	return rc;
+}
+
+static int auxdisplay_serlcd_custom_character_set(const struct device *dev,
+						  struct auxdisplay_character *character)
+{
+	const struct auxdisplay_serlcd_config *config = dev->config;
+
+	/*
+	 * only indexes 0..7 are supported
+	 */
+	const uint8_t char_index = character->index;
+
+	if (char_index > 7) {
+		return -EINVAL;
+	}
+
+	/*
+	 * custom characters are accessible via char codes 0x08..0x0f
+	 */
+	character->character_code = 0x08 | char_index;
+
+	int rc = auxdisplay_serlcd_send_command(dev, SERLCD_COMMAND_SET_CUSTOM_CHAR + char_index);
+
```

Comment:
nit: usually good practice to keep the check next to line that sets the return value, drop this blank line

Reply:
IIRC the compliance workflow demanded this extra line, but I will split definition and declaration of rc, so it complies and is directly above.